janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/computer/FulgoraGraphComputer.java

```diff
@@ -162,193 +179,241 @@ public GraphComputer mapReduce(final MapReduce mapReduce) {
-                        } finally {
-                            memory.incrIteration();
-                        }
+        Map<MapReduce, FulgoraMapEmitter> mapJobs = collectMapJobs();
+        executeMapJobs(mapJobs);
+
+        Graph resultgraph = writeMutatedPropertiesBackIntoGraph();
+        // update runtime and return the newly computed graph
+        this.memory.setRuntime(System.currentTimeMillis() - time);
+        this.memory.complete();
+        return new DefaultComputerResult(resultgraph, this.memory);
+    }
+
+    private void executeVertexProgram() {
+        if (null == vertexProgram) return;
+
+        vertexMemory = new FulgoraVertexMemory(expectedNumVertices, graph.getIDManager(), vertexProgram);
+        vertexProgram.setup(memory);
+
+        try (VertexProgramScanJob.Executor job = VertexProgramScanJob.getVertexProgramScanJob(graph, memory, vertexMemory, vertexProgram)) {
+            for (int iteration = 1; ; iteration++) {
+                memory.completeSubRound();
+                executeIterationOfJob(job, iteration);
+                memory.completeSubRound();
+                try {
+                    if (this.vertexProgram.terminate(this.memory)) {
+                        break;
                     }
+                } finally {
+                    memory.incrIteration();
                 }
             }
+        }
+    }
 
-            // ##### Execute map-reduce jobs
-            // Collect map jobs
-            Map<MapReduce, FulgoraMapEmitter> mapJobs = new HashMap<>(mapReduces.size());
-            for (MapReduce mapReduce : mapReduces) {
-                if (mapReduce.doStage(MapReduce.Stage.MAP)) {
-                    FulgoraMapEmitter mapEmitter = new FulgoraMapEmitter<>(mapReduce.doStage(MapReduce.Stage.REDUCE));
-                    mapJobs.put(mapReduce, mapEmitter);
-                }
+    private void executeIterationOfJob(VertexProgramScanJob.Executor job, int iteration) {
+        initializeVertexMemoryForIteration();
+        StandardScanner.Builder scanBuilder = createScanBuilderForJob(job, iteration);
+        PartitionedVertexProgramExecutor programExecutor = new PartitionedVertexProgramExecutor(graph, memory, vertexMemory, vertexProgram);
+        try {
+            //Iterates over all vertices and computes the vertex program on all non-partitioned vertices. For partitioned ones, the data is aggregated
+            ScanMetrics jobResult = executeOnNonPartitionedVertices(iteration, scanBuilder);
+
+            executeOnPartitionedVertices(iteration, programExecutor, jobResult);
+        } catch (Exception e) {
+            throw new JanusGraphException(e);
+        }
+
+        vertexMemory.completeIteration();
+    }
+
+    private void initializeVertexMemoryForIteration() {
+        if (vertexProgram instanceof ShortestPathVertexProgram) {
+            HashSet<MessageScope> locals = new HashSet<>(Collections.singletonList(MessageScope.Local.of(__::bothE)));
```

Comment:
(nitpick)
You are creating a singleton List and then HashSet here. I think singleton list is redundant here. You can use something like this:
```
vertexMemory.nextIteration(new HashSet<MessageScope>(){{
    add(MessageScope.Local.of(__::bothE));
    add(MessageScope.Global.instance());
}});
```
or 
```
HashSet<MessageScope> locals = new HashSet<>();
locals.add(MessageScope.Local.of(__::bothE));
locals.add(MessageScope.Global.instance());
vertexMemory.nextIteration(locals);
```